`storage`: change `has_tx_data_batch` condition

WillemKauf opened the pull request:
Quick followup to https://github.com/redpanda-data/redpanda/pull/28805.

On second thought, we need to capture the presence of both transactional data raft batches as well as `tx_fence` batches here when computing an MCCO equivalent (which tells us the offset below which control batches are removable).

Otherwise, we are not checking our condition stringently enough when determining what control batches are removable.

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

WillemKauf committed:
`storage`: change `has_tx_data_batch` condition

On second thought, we need to capture the presence of both transactional
data raft batches as well as `tx_fence` batches here when computing
an MCCO equivalent (which tells us the offset below which control batches
are removable).

bashtanov commented on src/v/storage/index_state.h:
```diff
@@ -240,7 +240,8 @@ struct index_state
-    // data batches (i.e raft data batches with a transactional bit set).
+    // data batches (i.e raft data batches with a transactional bit set, as well
+    // as tx_fence markers).
     bool may_have_transaction_data_batches{true};
```
Too late to rename it? E.g. `may_have_nonterminal_transactional_data`?

WillemKauf replied:
Not too late to rename it... but I'm not sure what I'd prefer. Probably not `nonterminal_transactional_data`...

WillemKauf replied:
@bharathv is this worth renaming to something like `may_have_tx_data_or_fences`? Is the current name misleading now?

bharathv replied:
I think it’s worth renaming. It now applies to both control and data batches, but the current “has…data_batches” name suggests otherwise. Maybe has_transactional_batches?

bashtanov replied:
> Maybe has_transactional_batches?

the problem is it still excludes commit/abort batches (which we don't have a common name for)

bharathv replied:
right, but "nonterminal" seems like a new terminology, would prefer not to introduce that. can't think of any other than `data_or_fences`..

WillemKauf replied:
Changed to `may_have_transaction_data_or_fence_batches`. It's verbose but oh well.

bashtanov reviewed (dismissed):
LGTM bar a naming issue

WillemKauf committed:
`storage`: one more `segment_tx_flags` test

Tests compacting with `log_compaction_disable_tx_batch_removal` enabled
and ensures that the right flags are set per `segment`, then compacts
again with `log_compaction_disable_tx_batch_removal` disabled and
checks that all control batches and raft transactional bits are gone.

WillemKauf commented:
Force push to:

* Rename to `may_have_transaction_data_or_fence_batches`
* Add another unit test for `segment` flags.

WillemKauf commented:
Oops. bad force push. Reverted!

WillemKauf commented:
/ci-repeat 1

WillemKauf committed:
`storage`: respect `log_compaction_disable_tx_batch_removal`

This seems to have been forgotten when tx batch removal was
re-enabled in the tree.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--6c6593ce-33db-421f-bb3d-8946b8e7533d-->
<!--TEST RESULTS-->
<!--0206909d-bb0f-499a-ba5e-decd8b3d13f5-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77311">build#77311</a></summary>


| test_class                 | test_method                              | test_arguments                                                                                                                           | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                      |
|:---------------------------|:-----------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest      | test_migrated_topic_data_integrity       | {"params": {"cancellation": {"dir": "in", "stage": "executed"}, "include_groups": true, "transfer_leadership": true, "use_alias": true}} | integration | https://buildkite.com/redpanda/redpanda/builds/77311#019ae7cd-16c7-4f2b-bd93-d1ecc58871db | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_migrated_topic_data_integrity            |
| PartitionReassignmentsTest | test_reassignments_cancel                | null                                                                                                                                     | integration | https://buildkite.com/redpanda/redpanda/builds/77311#019ae7cd-16c4-4e13-bcc3-046965c77bce | FLAKY         | 15/21    | upstream reliability is '97.57575757575758'. current run reliability is '71.42857142857143'. drift is 26.14719 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionReassignmentsTest&test_method=test_reassignments_cancel                |
| PrefixTruncateRecoveryTest | test_prefix_truncate_recovery            | {"acks": -1, "start_empty": false}                                                                                                       | integration | https://buildkite.com/redpanda/redpanda/builds/77311#019ae7cd-16c4-4e13-bcc3-046965c77bce | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PrefixTruncateRecoveryTest&test_method=test_prefix_truncate_recovery            |
| TombstoneRemovalTest       | test_single_slow_follower                | null                                                                                                                                     | integration | https://buildkite.com/redpanda/redpanda/builds/77311#019ae7d3-98de-4f63-97dc-6bcdcdf92be1 | FLAKY         | 19/21    | upstream reliability is '99.69879518072288'. current run reliability is '90.47619047619048'. drift is 9.2226 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TombstoneRemovalTest&test_method=test_single_slow_follower                      |
| TxAtomicProduceConsumeTest | test_basic_tx_consumer_transform_produce | {"with_failures": true}                                                                                                                  | integration | https://buildkite.com/redpanda/redpanda/builds/77311#019ae7cd-16c1-43b3-9a65-66ffd903c955 | FLAKY         | 20/21    | upstream reliability is '98.48866498740554'. current run reliability is '95.23809523809523'. drift is 3.25057 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TxAtomicProduceConsumeTest&test_method=test_basic_tx_consumer_transform_produce |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77338">build#77338</a></summary>


| test_class              | test_method         | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                              |
|:------------------------|:--------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| MountUnmountIcebergTest | test_simple_remount | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/77338#019ae9fb-7375-47c1-935f-7c15cdc26f99 | FLAKY         | 15/21    | upstream reliability is '79.21478060046189'. current run reliability is '71.42857142857143'. drift is 7.78621 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount |
</details>
<!--/body-->

bashtanov commented on src/v/storage/tests/compaction_e2e_multinode_test.cc:
```diff
@@ -351,16 +351,145 @@ FIXTURE_TEST(segment_tx_flags, compaction_multinode_test) {
+    BOOST_REQUIRE_EQUAL(segments.size(), num_segments + 1);
+
+    for (const auto& segment : log->segments()) {
+        if (segment->has_self_compact_timestamp()) {
```
Don't we know how many segments we have compacted? I assume no merging has happened.
I find this condition a bit vague, as it won't catch a situation where no segments have been compacted.

WillemKauf replied:
Pushed to make these conditions explicit for each segment.
